File: go/interest-is-interesting/interest_is_interesting.go
package interest

// InterestRate returns the interest rate for the provided balance.
func InterestRate(balance float64) float32 {
	var rate float32
	switch {
	case balance < 0:
		rate = float32(3.213)
	case balance >= 0 && balance < 1000:
		rate = float32(0.5)
	case balance >= 1000 && balance < 5000:
		rate = float32(1.621)
	case balance >= 5000:
		rate = float32(2.475)
	}
	return rate
}

// Interest calculates the interest for the provided balance.
func Interest(balance float64) float64 {
	rate := InterestRate(balance)

	return float64(rate/100) * balance
}

// AnnualBalanceUpdate calculates the annual balance update, taking into account the interest rate.
func AnnualBalanceUpdate(balance float64) float64 {
	interset := Interest(balance)

	return balance + interset
}

// YearsBeforeDesiredBalance calculates the minimum number of years required to reach the desired balance.
func YearsBeforeDesiredBalance(balance, targetBalance float64) int {
	var years int
	for balance < targetBalance {
		balance = AnnualBalanceUpdate(balance)
		years++
	}
	return years
}

File: cpp/interest-is-interesting/interest_is_interesting.cpp
// interest_rate returns the interest rate for the provided balance.
double interest_rate(double balance) {
  if (balance < 0) {
    return 3.213;
  } else if (balance >= 0 && balance < 1000) {
    return 0.5;
  } else if (balance >= 1000 && balance < 5000) {
    return 1.621;
  } else {
    return 2.475;
  }
}

// yearly_interest calculates the yearly interest for the provided balance.
double yearly_interest(double balance) {
  double rate = interest_rate(balance);
  return (rate / 100.0) * balance;
}

// annual_balance_update calculates the annual balance update, taking into
// account the interest rate.
double annual_balance_update(double balance) {
  double interest = yearly_interest(balance);
  return interest + balance;
}

// years_until_desired_balance calculates the minimum number of years required
// to reach the desired balance.
int years_until_desired_balance(double balance, double target_balance) {
  int years = 0;
  double balance_copy = balance;

  while (balance_copy < target_balance) {
    balance_copy = annual_balance_update(balance_copy);
    ++years;
  }

  return years;
}
